modules/enterprise/server/jar/src/main/java/org/rhq/enterprise/server/resource/ResourceManagerBean.java

```diff
@@ -783,6 +789,62 @@ private boolean uninventoryResourceBulkDeleteAsyncWork(Subject overlord, int res
                 resourceIds);
         }
 
+        // If this resource had packages installed, remove their version and bits
+        if (installedPackages.size() > 0) {
+            hasErrors |= cleanOrphanedPackages(overlord, installedPackages);
+        }
+
+        return hasErrors;
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<Integer> getInstalledPackages(List<Integer> resourceIds) {
+
+        List<Integer> installed_packages = new ArrayList<Integer>();
+
+        try {
+            installed_packages = entityManager.createNamedQuery(PackageVersion.QUERY_FIND_ID_PACKAGES_BY_RESOURCE_ID)
+                         .setParameter("resourceIds", resourceIds)
+                .getResultList();
+        } catch (Exception e) {
+            LOG.warn("May not have been able get installed packages for" + resourceIds
+                + ". This is unlikely to be a problem.", e);
+            return installed_packages;
+        }
+
+        return installed_packages;
+    }
+
+    @TransactionAttribute(TransactionAttributeType.REQUIRES_NEW)
+    private boolean cleanOrphanedPackages(Subject subject, List<Integer> packages) {
+
+        boolean hasErrors = false;
+
+        if (!authorizationManager.isOverlord(subject)) {
+            throw new IllegalArgumentException("Only the overlord can execute arbitrary named query strings");
+        }
+
+        String[] namedQueriesToExecute = new String[] { //
+        PackageVersion.DELETE_BY_PKG_IF_NO_CONTENT_SOURCES_OR_REPOS, //
+            PackageBits.DELETE_IF_NO_PACKAGE_VERSION };
+
+        for (String namedQueryToExecute : namedQueriesToExecute) {
+            try {
+                Query namedQuery = entityManager.createNamedQuery(namedQueryToExecute);
+                if (namedQueryToExecute.equals(PackageVersion.DELETE_BY_PKG_IF_NO_CONTENT_SOURCES_OR_REPOS)) {
+                    namedQuery.setParameter("packageIds", packages);
```

Comment:
I think this will fail on Oracle if there are more than 1000 packages.  Oracle has an IN clause limit of 1000. The same could be true of getInstalledPackages(resourceIds) but actually it is OK because this logic is invoked on a single resource.  So, the limit is not an issue for getInstalledPackages and I doubt a single resource has that many packages.